Use named class for UndoPair.

gkossakowski opened the pull request:
Use specific, named class for UndoPair instead of generic Tuple2. This
makes analysis of heap dumps much easier because profilers let you inspect
memory consumption on per-class basis.

gkossakowski commented:
Review by @retronym

retronym commented on src/reflect/scala/reflect/internal/tpe/TypeConstraints.scala:
```diff
@@ -16,8 +16,9 @@ private[internal] trait TypeConstraints {
   private lazy val _undoLog = new UndoLog
   def undoLog = _undoLog
 
+  final case class UndoPair(tv: TypeVar, tConstraint: TypeConstraint)
```
Will this carry around an outer pointer? I think it doesn't due to the fact it is final (which is special cased to allow slim nested classes).

gkossakowski replied:
It does carry an outer pointer, good catch:

```
Greks-MBP:build grek (undoPairs)$ javap -private './quick/classes/reflect/scala/reflect/internal/tpe/TypeConstraints$UndoPair'
Compiled from "TypeConstraints.scala"
public final class scala.reflect.internal.tpe.TypeConstraints$UndoPair extends java.lang.Object implements scala.Product,scala.Serializable{
    private final scala.reflect.internal.Types$TypeVar tv;
    private final scala.reflect.internal.tpe.TypeConstraints$TypeConstraint tConstraint;
    private final scala.reflect.internal.SymbolTable $outer;
```

I'll rework this PR.

gkossakowski committed:
Use named class for UndoPair.

Use specific, named class for UndoPair instead of generic Tuple2. This
makes analysis of heap dumps much easier because profilers let you inspect
memory consumption on per-class basis.

The UndoPair case class is defined in companion object to not hold an
outer pointer reference.

gkossakowski commented:
I moved UndoPair to companion object. It doesn't hold an outer pointer reference anymore.

retronym commented:
LGTM
